pkg/cluster/clusterserviceprincipal.go

```diff
@@ -77,47 +78,64 @@ func (m *manager) createOrUpdateClusterServicePrincipalRBAC(ctx context.Context)
 	return nil
 }
 
-func (m *manager) updateAROSecret(ctx context.Context) error {
+// servicePrincipalUpdated checks whether the CSP has been updated by comparing the cluster doc's
+// ServicePrincipalProfile to the contents of the kube-system/azure-cloud-provider Secret. If the CSP
+// has changed, it returns true along with a new corev1.Secret to use to update the Secret to match
+// what's in the cluster doc.
+func (m *manager) servicePrincipalUpdated(ctx context.Context) (bool, *corev1.Secret, error) {
 	var changed bool
 	spp := m.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
-	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		//data:
-		// cloud-config: <base64 map[string]string with keys 'aadClientId' and 'aadClientSecret'>
-		secret, err := m.kubernetescli.CoreV1().Secrets("kube-system").Get(ctx, "azure-cloud-provider", metav1.GetOptions{})
-		if err != nil {
-			if kerrors.IsNotFound(err) { // we are not in control if secret is not present
-				return nil
-			}
-			return err
+	//data:
+	// cloud-config: <base64 map[string]string with keys 'aadClientId' and 'aadClientSecret'>
+	secret, err := m.kubernetescli.CoreV1().Secrets("kube-system").Get(ctx, "azure-cloud-provider", metav1.GetOptions{})
+	if err != nil {
+		if kerrors.IsNotFound(err) { // we are not in control if secret is not present
+			return false, nil, nil
 		}
+		return false, nil, err
+	}
 
-		var cf map[string]interface{}
-		if secret != nil && secret.Data != nil {
-			err = yaml.Unmarshal(secret.Data["cloud-config"], &cf)
-			if err != nil {
-				return err
-			}
-			if val, ok := cf["aadClientId"].(string); ok {
-				if val != spp.ClientID {
-					cf["aadClientId"] = spp.ClientID
-					changed = true
-				}
+	var cf map[string]interface{}
+	if secret != nil && secret.Data != nil {
+		err = yaml.Unmarshal(secret.Data["cloud-config"], &cf)
+		if err != nil {
+			return false, nil, err
+		}
+		if val, ok := cf["aadClientId"].(string); ok {
+			if val != spp.ClientID {
+				cf["aadClientId"] = spp.ClientID
+				changed = true
 			}
-			if val, ok := cf["aadClientSecret"].(string); ok {
-				if val != string(spp.ClientSecret) {
-					cf["aadClientSecret"] = spp.ClientSecret
-					changed = true
-				}
+		}
+		if val, ok := cf["aadClientSecret"].(string); ok {
+			if val != string(spp.ClientSecret) {
+				cf["aadClientSecret"] = spp.ClientSecret
+				changed = true
 			}
 		}
+	}
 
-		if changed {
-			data, err := yaml.Marshal(cf)
-			if err != nil {
-				return err
-			}
-			secret.Data["cloud-config"] = data
+	if changed {
+		data, err := yaml.Marshal(cf)
+		if err != nil {
+			return false, nil, err
+		}
+		secret.Data["cloud-config"] = data
+		return true, secret, nil
+	}
+
+	return false, nil, nil
```

Comment:
fwiw: if we always return `true` with an object and `false` with nil, we don't need the boolean.